pkg/sign/signer.go

```diff
@@ -15,29 +15,218 @@
+}
+
+type publicKey struct {
+	Algorithm string `json:"algorithm"`
+	Content   string `json:"content"`
+}
+
+type fulcioResponse struct {
+	SCT signedCertificateEmbeddedSct `json:"signedCertificateEmbeddedSct"`
+}
+
+type signedCertificateEmbeddedSct struct {
+	Chain chain `json:"chain"`
+}
+
+type chain struct {
+	Certificates []string `json:"certificates"`
+}
+
+func (f *Fulcio) Sign(content Content) (*protobundle.Bundle, error) {
+	// Get JWT from identity token
+	tokenParts := strings.Split(f.options.IdentityToken, ".")
+	if len(tokenParts) < 2 {
+		return nil, errors.New("Unable to get subject from identity token")
+	}
+
+	jwtString, err := base64.RawURLEncoding.DecodeString(tokenParts[1])
+	if err != nil {
+		return nil, err
+	}
+
+	var jwt jsonWebToken
+	err = json.Unmarshal([]byte(jwtString), &jwt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate ephemeral keypair
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	pubKeyBytes, err := x509.MarshalPKIXPublicKey(privateKey.Public())
+	if err != nil {
+		return nil, err
+	}
+
+	pemBlock := pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: pubKeyBytes,
+	}
+
+	var pemBytes bytes.Buffer
+	if err = pem.Encode(&pemBytes, &pemBlock); err != nil {
+		return nil, err
+	}
+
+	// Sign JWT subject with ephemeral key for proof of possession
+	subjectDigest := sha256.Sum256([]byte(jwt.Sub))
+	subjectSignature, err := privateKey.Sign(rand.Reader, subjectDigest[:], nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make Fulcio certificate request
+	certRequest := fulcioCertRequest{
+		Credentials: identityToken{
+			OIDCIdentityToken: f.options.IdentityToken,
+		},
+		PublicKeyRequest: publicKeyRequest{
+			PublicKey: publicKey{
+				Algorithm: "ECDSA",
+				Content:   pemBytes.String(),
+			},
+			ProofOfPossession: base64.StdEncoding.EncodeToString(subjectSignature),
+		},
+	}
 
-// func (f *Fulcio) Sign(data []byte) (*protobundle.Bundle, error) {}
+	requestJSON, err := json.Marshal(&certRequest)
+	if err != nil {
+		return nil, err
+	}
+	requestBytes := bytes.NewBuffer(requestJSON)
+
+	var client http.Client
+	if f.options.Timeout != 0 {
+		client.Timeout = f.options.Timeout
+	}
+
+	request, err := http.NewRequest("POST", f.options.BaseURL+"/api/v2/signingCert", requestBytes)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add("Authorization", "Bearer "+f.options.IdentityToken)
+	request.Header.Add("Content-Type", "application/json")
+	request.Header.Add("User-Agent", constructUserAgent(f.options.LibraryVersion))
+
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("Fulcio returned %d: %s", response.StatusCode, string(body))
+	}
+
+	// Assemble bundle from Fulcio response
+	var fulcioResp fulcioResponse
+	err = json.Unmarshal(body, &fulcioResp)
```

Comment:
Does this work? I thought oneOf and json unmarshalling don't work well together.

In the WIP client, I was using protojson - https://gist.github.com/haydentherapper/86274b9211685a34b4da6b16725870a9#file-fulcio-client-go-L186-L189

Reply:
This works as the structure used is defined to match JSON serialization returned. I've been running similar code for another project.

Reply:
Oh good! I think I ran into this because I was trying to handle unmarshalling either the detached or the embedded. I really just need to mark the former as deprecated...